pwndbg/commands/xinfo.py

```diff
@@ -63,17 +63,38 @@ def xinfo_mmap_file(page, addr):
     file_name = page.objfile
     objpages = filter(lambda p: p.objfile == file_name, pwndbg.vmmap.get())
     first = sorted(objpages, key = lambda p: p.vaddr)[0]
+
+    # print offset from ELF base load address
     rva = addr - first.vaddr
+    print_line("File (Base)", addr, first.vaddr, rva, "+")
+
+    # find possible LOAD segments that designate memory and file backings
+    containing_loads = [seg for seg in pwndbg.elf.get_containing_segments(file_name, first.vaddr, addr)
+                        if seg['p_type'] == 'PT_LOAD']
+
+    for segment in containing_loads:
+        if segment['p_type'] == 'PT_LOAD' and addr < segment['x_vaddr_mem_end']:
+            offset = addr - segment['p_vaddr']
+            print_line('File (Segment)', addr, segment['p_vaddr'], offset, '+')
+            break
+
+    for segment in containing_loads:
+        if segment['p_type'] == 'PT_LOAD' and addr < segment['x_vaddr_file_end']:
+            file_offset = segment['p_offset'] + (addr - segment['p_vaddr'])
+            print_line("File (Disk)", addr, file_name, file_offset, "+")
```

Comment:
There is one whitespace too much on one of my systems, idk why; I am fine to leave it as it is as it is not that important but if u have time to look at it, it would be nice

```
  Offset information:
         Mapped Area 0x4048c5 = 0x400000 + 0x48c5
         File (Base) 0x4048c5 = 0x400000 + 0x48c5
      File (Segment) 0x4048c5 = 0x400000 + 0x48c5
         File (Disk) 0x4048c5 = /bin/ls   + 0x48c5
```

Reply:
Cannot reproduce.

Have tried with ELFs with short and long names, and with small (0x400000) and large (0x555555554000) load addresses. All produce consistent results: Space padding up to the length of the address above, or else just a single space before the "+".

The code in `print_line` for formatting the filename looks sound, and seems to work consistently across py2 and 3 on my system.

Could you try doing `file_name.strip()` on systems that show this bug? Cannot see how it would be the case that some spaces are hiding there without breaking stuff, but currently have no other leads.

Hmm, also not entirely sure about py2 vs 3 'type()' semantics - could this line

```
M.get(first) if type(first) is not str else first.ljust(len(hex(addr)))
```

sometimes result in `M.get()` rendering `first` instead of `first.ljust(len(hex(addr)))` for whatever type `page.objfile` has?

Reply:
It is better to do `not isinstance(first, str)`

Reply:
The problem with formatting is on Py2 as `hex(addr)` might return an `L` at the end:
```
(Pdb) hex(addr)
'0x4048c5L'
```

Reply:
I have fixed both things (`isinstance` and `.rstrip('L')`).